DPO Recipe Doc (#2091)
pytorch/torchtune: docs/source/recipes/dpo.rst

.. _dpo_recipe_label:

====================================
Direct Preference Optimization
====================================

This recipe supports several `Direct Preference Optimization <https://arxiv.org/abs/2305.18290>`_ (DPO)-style fine-tuning techniques.
These techniques aim to steer (or `align <https://en.wikipedia.org/wiki/AI_alignment>`_) a model towards some desirable behaviours.
For example, a common goal is to train language models to produce safe and honest outputs,
or to be `helpful and harmless <https://arxiv.org/abs/2204.05862>`_.

To see the best results when using this recipe, it may be helpful to first fine-tune your model with using supervised fine-tuning to ensure your model is
on-distribution for the domain you're interested in. To do this, check out our other fine-tuning recipes in the :ref:`recipe overview <recipes_overview_label>` which
support a variety of SFT paradigms.

> We provide configs which you can get up and running quickly. Here is an example with Llama 3.1 8B:

Reviewer: ```suggestion
After supervised fine-tuning, here is an example of DPO with Llama 3.1 8B:
```

.. note::

    You may need to be granted access to the Llama model you're interested in. See
    :ref:`here <download_llama_label>` for details on accessing gated repositories.

.. code-block:: bash

    tune download meta-llama/Meta-Llama-3.1-8B-Instruct \

>     --output-dir /tmp/Meta-Llama-3.1-8B-Instruct \

Reviewer: You can remove this.

    --ignore-patterns "original/consolidated.00.pth"
    --HF_TOKEN <HF_TOKEN>

    # run on a single device
    tune run lora_dpo_single_device --config llama3_1/8B_lora_dpo_single_device

    # run on two gpus

>     tune run --nnodes 1 --nproc_per_node 2 lora_dpo_distributed --config llama3_1/8B_lora_dpo

Reviewer: No need to specify nodes
